lib/core/jobs/consumer.rb

```diff
@@ -0,0 +1,120 @@
+
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+# frozen_string_literal: true
+
+module Core
+  module Jobs
+    class Consumer
+      def initialize(poll_interval: 3, termination_timeout: 60, min_threads: 1, max_threads: 5, max_queue: 100, idle_time: 5)
+        @poll_interval = poll_interval
+        @termination_timeout = termination_timeout
+        @min_threads = min_threads
+        @max_threads = max_threads
+        @max_queue = max_queue
+        @idle_time = idle_time
+
+        @running = Concurrent::AtomicBoolean.new(false)
+
+        Kernel.at_exit { shutdown! }
+      end
+
+      def subscribe!(index_name:)
+        @index_name = index_name
+
+        start_loop!
+      end
+
+      def running?
+        # @TODO check if a loop thread is alive
+        pool.running? && @running.true?
+      end
+
+      def shutdown!
+        Utility::Logger.info("Shutting down consumer for #{@index_name} index")
+        @running.make_false
+        pool.shutdown
+        pool.wait_for_termination(@termination_timeout)
+        # reset pool
+        @pool = nil
+      end
+
+      private
+
+      def start_loop!
+        Utility::Logger.info("Starting a new consumer for #{@index_name} index")
+
+        Thread.new do
```

Comment:
Can the thread actually exit and die somehow? I see you rescue StandardError, but Exceptions can cause it to die and stop reporting. While Exceptions probably indicate that connector won't actually run at all, it might be valuable to keep receiving errors and trying again so that user in Cloud would be able to see errors in their deployment instead of seeing it as running but not producing proper logs.

Reply:
I will follow up with a different PR to address this problem. 

`TimerTask` from ruby-concurrent is a good option.